[Name lookup] Don't look at a type's members when checking inherited types

DougGregor opened the pull request:
Teach unqualified name lookup not to look at the members of a nominal type or extension therefore when performing lookup in the inheritance clause of nominal type or extension. This was previously implemented in the type checker, but sink this logic down into `UnqualifiedLookup`, removing it from the type checker.

Built on top of @jrose-apple's excellent start.

jrose-apple committed:
[AST] Drop redundant ProtocolDecl::TrailingWhereClause

There's already a field for this in GenericContext, which ProtocolDecl
indirectly inherits. Protocols may have slightly different treatment
of their where-clauses, but not enough to justify a separate field.

No functionality change.

jrose-apple committed:
[AST] Don't look at a type's members when checking inherited types

Without this, we were firing off way more InheritedDeclsReferenced-
Requests than were actually needed. This drops it way down in the
profile, for what I /think/ is minor wins in type-checking the
stdlib...but it's hard to tell, since the time varies from run to run.

DougGregor committed:
[Name lookup] Factor out the check controlling member lookup into a nominal.

DougGregor committed:
[Name lookup] Generalize member lookup check for extensions as well.

We only look for members within the extension’s type when we are in the
body or a protocol extension’s where clause.

DougGregor committed:
[Type checker] Eliminate TypeResolverContext::GenericSignature.

The name-lookup behavior that avoids looking for members of a nominal type
or extension therefore when resolving the inheritance clause is now
handled by UnqualifiedLookup, so remove it from the type checker itself.

Fixes rdar://problem/39130543.

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -3192,7 +3192,8 @@ Type TypeBase::getSuperclassForDecl(const ClassDecl *baseClass,
   }
-  llvm_unreachable("no inheritance relationship between given classes");
+
+  return ErrorType::get(this);
```
How did we end up here? Name lookup shouldn't be finding members of unrelated classes...

DougGregor replied:
I suspect that Decl-based name lookup is finding something via a superclass (decl), but the types are invalid so we don't see them in the superclass (Type) chain. I can try to narrow this.

DougGregor replied:
See https://github.com/apple/swift/pull/18720/commits/50763fc641a29cb178fa7e6f60ac6c81bb77fad5

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -244,25 +244,20 @@ void TypeChecker::checkInheritanceClause(Decl *decl,
-      options = TypeResolutionOptions(TypeResolverContext::GenericSignature);
     } else if (!DC->isModuleScopeContext()) {
       // Skip the generic parameter's context entirely.
       DC = DC->getParent();
```
So the only other case is a SubscriptDecl, and it seems like this whole series of if/else if statements could just be removed? Just set `DC = decl->getDeclContext()` and be done with it.

DougGregor replied:
Ah, I was planning to rewrite this in terms of `InheritedTypeRequest`, because this whole pile of code is redundant.

slavapestov commented on validation-test/compiler_crashers_2_fixed/0153-rdar39130543.swift:
```diff
@@ -1,5 +1,5 @@
-// RUN: not --crash %target-swift-frontend %s -emit-ir
-// REQUIRES: asserts
+// RUN: %target-swift-frontend %s -emit-ir
```
This is not-crashing for the wrong reasons still :)

DougGregor replied:
Ahh, right. This one. Thanks for the reminder

xedin commented on lib/AST/NameLookup.cpp:
```diff
@@ -579,6 +579,41 @@ SelfBoundsFromWhereClauseRequest::evaluate(Evaluator &evaluator,
+
+  // When we have no source-location information, we have to perform member
+  // lookuo.
+  if (loc.isInvalid() || decl->getBraces().isInvalid())
```
nit: `lookuo` -> `lookup` :)

DougGregor replied:
Got it, thanks

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Those debug-build improvements look nice. Yay.

DougGregor committed:
[Name lookup] Fix typo in comment. NFC

DougGregor committed:
[Type checker] Fix DeclContext computation for inherited type lookup.

DougGregor committed:
[GSB] Make requirement-signature computation more robust for nested protocols.

Add all of the generic parameters from outer contexts as well. This only
occurs in ill-formed code, but maintains a GenericSignatureBuilder
invariant that it knows about all of the generic parameters in play.

DougGregor committed:
[AST] Re-establish assertion about inheritance relationship in getSuperclassForDecl()

When we fail to find a particular superclass type of a type in ill-formed
code, make sure that we were right to look here at all. This re-establishes
a check I had recently weakened.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge
